qadence2_platforms/backends/_base_analog/functions.py

```diff
@@ -0,0 +1,212 @@
+        """
+        Convert an expression into a native QuTiP object. A simple symbol,
+        a quantum operator, and an operation (addition, multiplication or
+        kron tensor) are valid objects.
+
+        Args:
+            num_qubits (int): the number of qubits to create the qutip object to
+            op (InputType): the input expression. Any qadence2-expressions expression
+                compatible object, with the same methods
+
+        Returns:
+            A QuTiP object with the Hilbert space compatible with `num_qubits`
+        """
+
+        if op.is_symbol is True:
+            symbol: str = cast(str, op.args[0])
+            return cls.operators_mapping[symbol]
+
+        op_arg: qutip.Qobj
+
+        if op.is_quantum_operator is True:
+            sub_num_qubits: int = cast(Support, op.args[1]).max_index + 1
+
+            if sub_num_qubits < num_qubits:
+                op_arg = cls._compl_tensor_op(num_qubits, op)
+
+            else:
+                arg: InputType = cast(InputType, op.args[0])
+                op_arg = cls._get_op(num_qubits, arg)
+
+            return op_arg
+
+        ops: list[qutip.Qobj] = []
+        args: Iterable = cast(Iterable, op.args)
+
+        if op.is_addition is True:
+
+            for arg in args:
+                ops.append(cls._arith_tensor_op(num_qubits, arg))
+
+            return reduce(lambda a, b: a + b, ops)
+
+        if op.is_multiplication is True:
+
+            for arg in args:
+                ops.append(cls._arith_tensor_op(num_qubits, arg))
+
+            return reduce(lambda a, b: a * b, ops)
+
+        if op.is_kronecker_product is True:
+            return cls._compl_tensor_op(num_qubits, op)
+
+        raise NotImplementedError(
+            f"could not retrieve the expression {op} ({type(op)}) from the observables"
+        )
+
+    @classmethod
+    def _iterate_over_obs(cls, n_qubits: int, op: Iterable | InputType) -> list[qutip.Qobj]:
+        if isinstance(op, Iterable):
+            return [cls._get_op(n_qubits, arg) for arg in op]
+
+        args: Iterable = cast(Iterable, op.args)
+        return [cls._get_op(n_qubits, cast(InputType, arg)) for arg in args]
+
+    @classmethod
+    def build(cls, num_qubits: int, observables: list[InputType] | InputType) -> list[qutip.Qobj]:
+        """
+        Parses an input expression or list of expressions into a native QuTiP object.
+
+        Args:
+            num_qubits (int): the number of qubits to create the qutip object to
+            observables (InputType): the input expression. Any qadence2-expressions
+                expression compatible object, with the same methods
+
+        Returns:
+            A QuTiP object with the Hilbert space compatible with `num_qubits`
+
+        Returns:
+        """
+        if not isinstance(observables, list):
+            return [cls._get_op(num_qubits, observables)]
+        return cls._iterate_over_obs(num_qubits, observables)
```

Comment:
Shouldn't this be rewritten into a set of standalone functions? The class `BaseQuTipObservablesParser` only has classmethods.

Reply:
no, because they work as a whole. It is to be thought as a context-giving class